### ISSUE TITLE
Can't connect to BLE / Carly OBDII Adapter

### ISSUE THREAD

**Issue by derkrasseleo** [osmandapp/OsmAnd#21441]
### Description

I can't connect my Carly OBDII Adapter. It works in KartaView and CarScanner.

### Steps to reproduce

- Go to vehicle metrics settings
- Click on connect
- Try to select a BLE device

### Actual result

The connect button does nothing when I click on it.

### Expected result

Should connect and show data.

### Your Environment (required)

```
OsmAnd Version: 4.9.7
Android/iOS version: Android 13 (GrapheneOS)
Device model: Pixel 4a
Crash-Logs: -
```

**Comment by vshcherb** [comment]
Did you try to pair device in Bluetooth Settings first? We would need to have logs to see what's going on (Help -> Logs), you can attach here as zip file

**Comment by derkrasseleo** [comment]
> Did you try to pair device in Bluetooth Settings first? We would need to have logs to see what's going on (Help -> Logs), you can attach here as zip file

Yes, and it doesn't, but I think the problem is that BLE devices generally don't pair like normal Bluetooth devices do..

**Comment by dmpr0** [comment]
The issue is likely that the scanner is specifically designed to work only with the Carly app. We are aware of a similar issue with OBDeleven.

**Comment by derkrasseleo** [comment]
> The issue is likely that the scanner is specifically designed to work only with the Carly app. We are aware of a similar issue with OBDeleven.

As I said, it works with KartaView and CarScanner..

**Comment by derkrasseleo** [comment]
> Did you try to pair device in Bluetooth Settings first? We would need to have logs to see what's going on (Help -> Logs), you can attach here as zip file
Here's the log file, I can't find anything related to the plugin:
[logcat(2).log](https://github.com/user-attachments/files/17921989/logcat.2.log)

I also noticed that I can't see the other buttons (settings, rename, forget) when trying to connect to the device

**Comment by vshcherb** [comment]
Expensive devices usually have own protocol or custom way to communicate. So far we support simple Bluetooth devices, these devices usually needs to be paired first as Bluetooth device. 

We will add more logs to understand what went wrong and how it could be implemented. Unfortunately CarScanner is not opensource to understand what's specific Carly has.

**Comment by derkrasseleo** [comment]
Thanks, let me know if I can provide more logs.

**Comment by vshcherb** [comment]
Yes looks like we need to get BLE-device for testing purposes and implement it cause Bluetooth differs from BLE.

**Comment by derkrasseleo** [comment]
> Yes looks like we need to get BLE-device for testing purposes and implement it cause Bluetooth differs from BLE.

That's what I suspected, maybe the implementation of BLE in KartaView helps:
https://github.com/kartaview/android/blob/master/app/src/main/java/com/telenav/osv/obd/BLEConnection.java

**Comment by Zirochkabila** [comment]
# Testing 1
`OsmAnd~ 5.2.0#5833m, Released: 2025-08-13`
`Vgate iCar Pro OBD2 Bluetooth 4.0 (BLE)`

Connection to the BLE sensor is checked - connection is successful, data is displayed.

@derkrasseleo  please check on your own device.

<details><summary>Details</summary>
<video src="https://github.com/user-attachments/assets/bc542f9c-67ec-44bd-8048-225257c4d7da"></video>
</details>

**Comment by derkrasseleo** [comment]
Did you compile it from this repo with Android Studio? Which branch did you use? I can't build it..
Could you send me the apk?

**Comment by Zirochkabila** [comment]
@derkrasseleo Of course. Please try this version of the Nightly Build https://download.osmand.net/latest-night-build/OsmAnd-default.apk

**Comment by derkrasseleo** [comment]
OsmAnd Nightly 5.2.0#5846m, released: 2025-08-14

Doesnt work for me, cant even connect to the Carly OBD Adapter, but every other known Bluetooth Device works somehow..

https://github.com/user-attachments/assets/a730a28a-7304-48c3-9f39-9ff825831548

**Comment by Zirochkabila** [comment]
@Corwin-Kh Please check on your own sensor and report the result.

**Comment by Corwin-Kh** [comment]
Hello, @derkrasseleo. Thanks for checking the build. As i can see from provided video, you attempting to connect to general bluetooth connection to device. It needs to be paired and really, some devices can be paired only with their own applications. In new build we added support to connect to devices via BLE - while searching for new device it should be listed in general list, and it's address should be preceeded by "BLE" mark. As this searching accurs in parralel, it can take more time to find ble devices. Try to wait additional time (in some cases it took about 30 seconds in my case) and if there is such item (there should be two Carly devices - one for BLE, one for bluetooth) and connect to ble. If this doesn't help we need you logs to investigate the problem

**Comment by derkrasseleo** [comment]
I just tried this build: OsmAnd Nightly 5.2.0#5891m, released: 2025-08-22

Same problem, the device doesn't want to connect, but I've also upgraded to Android 16 in the meantime, so there could be another issue.. 
I only got a short error message saying somthing like "pairing not possible", this is the screen After scanning for a few minutes, no BLE to be seen:

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/85d40bc9-888b-4e7c-9e30-a1971824edc6"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by derkrasseleo** [comment]
CarScanner still works with Android 16 though, and now I also got a system message that said: "pairing rejected" when trying to connect via Osmand

Here are slme interesting Things i found in the CarScanner log:

```BTLEDeviceID=00000000-0000-0000-0000-001a2201b4a7 BTLEDeviceName=Carly9DC986 BTLEInputID=0000ffe1-0000-1000-8000-00805f9b34fb BTLEOutputID=0000ffe1-0000-1000-8000-00805f9b34fb BTLEServiceID=0000ffe0-0000-1000-8000-00805f9b34fb```

**Comment by dmpr0** [comment]
## UI Review

- [x] Unable to connect to iOS Vlink (Vgate iCar Pro) on Android.

**Comment by DmitryAlexei** [comment]
- [x] migration for old devices

**Comment by Zirochkabila** [comment]
# Testing 2
`OsmAnd~ 5.2.0#6236m, Released: 2025-10-09`

Unable to connect to iOS Vlink (Vgate iCar Pro) on Android - **Still present**

<details><summary>Details</summary>
<video src="https://github.com/user-attachments/assets/0edbf226-814e-4378-9117-572443345a7e"></video>
</details>

**Comment by dmpr0** [comment]
## UI Review
Tested on `Google Pixel 6A` with `OsmAnd~ 5.3.0#6291m, дата випуску: 2025-10-17`

<table style="width:100%">
  <tr>
    <th>Android Vlink</th>
    <th>iOS Vlink</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/89a8b506-d60b-477a-b19b-e6e5d50fbb8a"  width="200" height="auto"/> </td>
    <td><img src="https://github.com/user-attachments/assets/a0eee715-ed4b-44a8-a460-4ee3079b0608"  width="200" height="auto"/> </td>
  </tr>
</table>   

<details>
    <summary>video</summary>
    <table style="width:100%">
      <tr>
        <th></th>
      </tr>
      <tr>
        <td><video src="https://github.com/user-attachments/assets/2837ec05-bdca-4359-82d7-d6e33c4d4a5a"  width="300" height="auto"/> </td>
      </tr>
    </table>
</details>
